src/Autofac/Core/Resolving/ResolveOperationBase.cs

```diff
@@ -246,43 +184,113 @@ public object GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, R
+                }
+
+                throw;
+            }
+            catch (DependencyResolutionException dependencyResolutionException)
+            {
+                if (DiagnosticSource.IsEnabled())
+                {
+                    DiagnosticSource.OperationFailure(this, dependencyResolutionException);
+                }
+
+                End(dependencyResolutionException);
+                throw;
+            }
+            catch (Exception exception)
+            {
+                End(exception);
+                if (DiagnosticSource.IsEnabled())
+                {
+                    DiagnosticSource.OperationFailure(this, exception);
+                }
+
+                throw new DependencyResolutionException(ResolveOperationResources.ExceptionDuringResolve, exception);
+            }
+            finally
+            {
+                ResetSuccessfulRequests();
+            }
+
+            End();
+
+            if (DiagnosticSource.IsEnabled())
+            {
+                DiagnosticSource.OperationSuccess(this, result);
+            }
+
+            return result;
+        }
+
         /// <summary>
         /// Basic pipeline invocation steps used when retrieving an instance. Isolated
         /// to enable it to be optionally surrounded with diagnostics.
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void InvokePipeline(ResolveRequest request, ResolveRequestContext requestContext)
+        private void InvokePipeline(ResolveRequest request, ResolveRequestContext requestContextContext)
         {
-            request.ResolvePipeline.Invoke(requestContext);
-            if (requestContext.Instance == null)
+            request.ResolvePipeline.Invoke(requestContextContext);
+            if (requestContextContext.Instance == null)
             {
                 throw new DependencyResolutionException(ResolveOperationResources.PipelineCompletedWithNoInstance);
             }
 
-            _successfulRequests.Add(requestContext);
+            _successfulRequests.Add(requestContextContext);
         }
 
         private void CompleteRequests()
         {
             var completed = _successfulRequests;
-            int count = completed.Count;
+            var count = completed.Count;
             var startPosition = _nextCompleteSuccessfulRequestStartPos;
             ResetSuccessfulRequests();
 
-            for (int i = startPosition; i < count; i++)
+            for (var i = startPosition; i < count; i++)
             {
-                completed[i].Complete();
+                completed[i].CompleteRequest();
             }
         }
 
-        private void ResetSuccessfulRequests()
-        {
-            _nextCompleteSuccessfulRequestStartPos = _successfulRequests.Count;
-        }
+        private void ResetSuccessfulRequests() => _nextCompleteSuccessfulRequestStartPos = _successfulRequests.Count;
```

Comment:
Not sure how I feel about expression-based methods that don't return things...style preference I think.